File: server/src/agents/researcher.js
import { createMessage, extractJson } from "./claude.js";
import { WEB_SEARCH_TOOL, webSearch } from "../tools/webSearch.js";

const RESEARCHER_SYSTEM = `You are the Researcher agent in a multi-agent job application pipeline.

Use the web_search tool to gather recent, factual information about the target company and role context.
Run multiple focused searches (overview/products, recent news, culture/mission) before you finish.

When you have enough information, respond with ONLY a JSON object (no markdown fences) in this exact shape:
{
  "companyName": string,
  "companyFacts": [string, ...],
  "roleRequirements": [string, ...],
  "cultureSignals": [string, ...],
  "sources": [string, ...],
  "researchNotes": string
}

Rules:
- companyFacts: 3-5 concrete facts grounded in search results
- roleRequirements: 4-6 requirements extracted from the job description
- cultureSignals: 2-3 culture/value signals
- sources: source domains from search results
- researchNotes: concise bullet-style notes for debugging
- Do not invent unsupported company facts`;

/**
 * @param {string} companyName
 * @param {string} jobDescription
 * @param {(event: import('./pipeline.js').TraceEvent) => void} emit
 */
export async function runResearcher(companyName, jobDescription, emit) {
  emit({
    type: "agent_start",
    agent: "researcher",
    message: `Starting web research for ${companyName}`,
    timestamp: new Date().toISOString(),
  });

  /** @type {import('@anthropic-ai/sdk').MessageParam[]} */
  const messages = [
    {
      role: "user",
      content: `Research this company for a job application.

Company: ${companyName}

Job description:
"""
${jobDescription}
"""

Use web_search as needed, then return the final JSON handoff object.`,
    },
  ];

  const searchLog = [];
  let structured = null;

  for (let turn = 0; turn < 8; turn += 1) {
    const response = await createMessage({
      system: RESEARCHER_SYSTEM,
      tools: [WEB_SEARCH_TOOL],
      messages,
      temperature: 0.2,
    });

    const toolUses = response.content.filter((block) => block.type === "tool_use");
    const textBlocks = response.content.filter((block) => block.type === "text");

    if (toolUses.length) {
      messages.push({ role: "assistant", content: response.content });

      /** @type {import('@anthropic-ai/sdk').ToolResultBlockParam[]} */
      const toolResults = [];

      for (const toolUse of toolUses) {
        if (toolUse.name !== "web_search") continue;
        const query =
          typeof toolUse.input === "object" &&
          toolUse.input &&
          "query" in toolUse.input &&
          typeof toolUse.input.query === "string"
            ? toolUse.input.query
            : `${companyName} company`;

        emit({
          type: "tool_call",
          agent: "researcher",
          message: `web_search("${query}")`,
          payload: { query },
          timestamp: new Date().toISOString(),
        });

        const bundle = await webSearch(query);
        searchLog.push(bundle);

        emit({
          type: "tool_result",
          agent: "researcher",
          message: `Found ${bundle.results.length} result(s) for "${query}"`,
          payload: bundle,
          timestamp: new Date().toISOString(),
        });

        toolResults.push({
          type: "tool_result",
          tool_use_id: toolUse.id,
          content: bundle.text,
        });
      }

      messages.push({ role: "user", content: toolResults });
      continue;
    }

    const text = textBlocks.map((block) => block.text).join("\n").trim();
    if (!text) {
      throw new Error("Researcher returned an empty response");
    }

    structured = extractJson(text);
    if (!structured.companyName) structured.companyName = companyName;
    if (!structured.sources?.length) {
      structured.sources = [
        ...new Set(searchLog.flatMap((entry) => entry.results.map((item) => item.url))),
      ]
        .map((url) => {
          try {
            return new URL(url).hostname.replace(/^www\./, "");
          } catch {
            return url;
          }
        })
        .filter(Boolean)
        .slice(0, 8);
    }

    emit({
      type: "handoff",
      agent: "researcher",
      message: "Structured JSON handoff ready for Writer",
      payload: structured,
      timestamp: new Date().toISOString(),
    });

    emit({
      type: "agent_complete",
      agent: "researcher",
      message: "Research complete",
      timestamp: new Date().toISOString(),
    });

    return { research: structured, searchLog };
  }

  throw new Error("Researcher exceeded maximum tool-calling turns");
}

File: backend/app/ai/researcher.py
from __future__ import annotations

from dataclasses import dataclass

from ddgs import DDGS

from backend.app.integrations.llm import Settings, get_settings
from backend.app.integrations.llm import chat, extract_json
from backend.app.ai.models import CompanyResearch

RESEARCH_SYSTEM = """You are a research assistant. Use the provided web search results to find recent, relevant
information about the given company: recent news, products, mission/values, and
anything notable in the last 6-12 months. Be factual and concise. Do not invent facts
not supported by the search results. If search results are thin, say so explicitly."""

HANDOFF_SYSTEM = """You convert raw research notes and a job description into a clean structured JSON
object for another AI agent to use when drafting a job application.

Return JSON with this exact shape:
{
  "companyName": string,
  "companyFacts": [string, ...],
  "roleRequirements": [string, ...],
  "cultureSignals": [string, ...],
  "sources": [string, ...]
}

Rules:
- companyFacts: 3-5 concrete, specific facts grounded in the research notes
- roleRequirements: 4-6 key requirements extracted from the job description
- cultureSignals: 2-3 signals about values/culture from research or JD
- sources: source names/domains from the research, or [] if none
- Do not invent unsupported company facts"""


@dataclass
class SearchBundle:
    text: str
    sources: list[str]


def web_search(company_name: str, *, max_results: int | None = None) -> SearchBundle:
    """Search the web for recent company information."""
    settings = get_settings()
    limit = max_results or settings.max_search_results
    queries = [
        f"{company_name} company overview products services",
        f"{company_name} recent news 2025 2026",
        f"{company_name} mission values culture",
    ]

    seen_urls: set[str] = set()
    lines: list[str] = []
    sources: list[str] = []

    with DDGS() as ddgs:
        for query in queries:
            try:
                results = list(ddgs.text(query, max_results=max(3, limit // 2)))
            except Exception as exc:  # noqa: BLE001 — keep pipeline resilient
                lines.append(f"[search error for '{query}']: {exc}")
                continue

            for item in results:
                url = (item.get("href") or item.get("link") or "").strip()
                title = (item.get("title") or "").strip()
                body = (item.get("body") or item.get("snippet") or "").strip()
                if not body:
                    continue
                if url and url in seen_urls:
                    continue
                if url:
                    seen_urls.add(url)
                    sources.append(_domain(url))
                lines.append(f"- {title}\n  URL: {url or 'n/a'}\n  {body}")

            if len(lines) >= limit * 2:
                break

    # Deduplicate source domains while preserving order
    unique_sources = list(dict.fromkeys(sources))
    text = "\n".join(lines) if lines else f"No web results found for {company_name}."
    return SearchBundle(text=text, sources=unique_sources)


def _domain(url: str) -> str:
    try:
        from urllib.parse import urlparse

        host = urlparse(url).netloc.lower()
        return host[4:] if host.startswith("www.") else host
    except Exception:  # noqa: BLE001
        return url


def summarize_search(company_name: str, search_text: str) -> str:
    """Step 1: turn raw search hits into concise research notes."""
    user = f"""Research this company for a job application: {company_name}.
Find recent news, main products/services, and culture or values signals.

Web search results:
\"\"\"
{search_text}
\"\"\"

Write concise factual research notes (bullet points). Cite source domains inline when possible."""
    return chat(
        [
            {"role": "system", "content": RESEARCH_SYSTEM},
            {"role": "user", "content": user},
        ],
        temperature=0.2,
    )


def structure_handoff(
    company_name: str,
    research_notes: str,
    job_description: str,
    sources: list[str] | None = None,
) -> CompanyResearch:
    """Step 2: convert notes + JD into structured JSON for the Writer."""
    source_hint = ", ".join(sources or []) or "(none)"
    user = f"""Raw research notes about the company:
\"\"\"
{research_notes}
\"\"\"

Known source domains from search: {source_hint}

Job description:
\"\"\"
{job_description}
\"\"\"

Company name: {company_name}

Return only the JSON object."""

    raw = chat(
        [
            {"role": "system", "content": HANDOFF_SYSTEM},
            {"role": "user", "content": user},
        ],
        temperature=0.1,
        json_mode=True,
    )
    data = extract_json(raw)
    if not data.get("companyName"):
        data["companyName"] = company_name
    if not data.get("sources") and sources:
        data["sources"] = sources[:8]
    return CompanyResearch.model_validate(data)


def research_company(
    company_name: str,
    job_description: str,
    *,
    settings: Settings | None = None,
) -> tuple[CompanyResearch, str]:
    """Full Researcher agent: web search → notes → structured handoff.

    Returns (structured_research, raw_notes).
    """
    _ = settings or get_settings()
    bundle = web_search(company_name)
    notes = summarize_search(company_name, bundle.text)
    structured = structure_handoff(
        company_name,
        notes,
        job_description,
        sources=bundle.sources,
    )
    return structured, notes
